12/umbraco-cms/reference/content-delivery-api/extension-points/custom-property-editors-support.md

```diff
@@ -0,0 +1,227 @@
+using Umbraco.Cms.Core.PropertyEditors;
+using Umbraco.Cms.Core.PropertyEditors.DeliveryApi;
+using Umbraco.Cms.Core.PublishedCache;
+
+namespace Umbraco.Docs.Samples;
+
+public class MyCustomPickerValueConverter : PropertyValueConverterBase, IDeliveryApiPropertyValueConverter
+{
+    private readonly IPublishedSnapshotAccessor _publishedSnapshotAccessor;
+    private readonly IApiContentRouteBuilder _apiContentRouteBuilder;
+
+    public MyCustomPickerValueConverter(
+        IPublishedSnapshotAccessor publishedSnapshotAccessor,
+        IApiContentRouteBuilder apiContentRouteBuilder)
+    {
+        _publishedSnapshotAccessor = publishedSnapshotAccessor;
+        _apiContentRouteBuilder = apiContentRouteBuilder;
+    }
+
+    public override bool IsConverter(IPublishedPropertyType propertyType)
+        => propertyType.EditorAlias.Equals("My.Custom.Picker");
+
+    public override Type GetPropertyValueType(IPublishedPropertyType propertyType)
+        => typeof(Guid?);
+
+    public PropertyCacheLevel GetDeliveryApiPropertyCacheLevel(IPublishedPropertyType propertyType)
+        => PropertyCacheLevel.Elements;
+
+    public Type GetDeliveryApiPropertyValueType(IPublishedPropertyType propertyType)
+        => typeof(DeliveryApiCustomPicker);
+
+    public object? ConvertIntermediateToDeliveryApiObject(
+        IPublishedElement owner,
+        IPublishedPropertyType propertyType,
+        PropertyCacheLevel referenceCacheLevel,
+        object? inter,
+        bool preview,
+        bool expanding)
+    {
+        if (inter is null)
+        {
+            return null;
+        }
+
+        return BuildDeliveryApiCustomPicker(inter, expanding);
+    }
+
+    private DeliveryApiCustomPicker? BuildDeliveryApiCustomPicker(object inter, bool expanding)
+    {
+        if (!_publishedSnapshotAccessor.TryGetPublishedSnapshot(out IPublishedSnapshot? publishedSnapshot) ||
+            publishedSnapshot?.Content is null)
+        {
+            return null;
+        }
+
+        if (!Guid.TryParse(inter as string, out Guid id))
+        {
+            return null;
+        }
+
+        return new DeliveryApiCustomPicker { Id = id };
+    }
+}
+
+public class DeliveryApiCustomPicker
+{
+    public Guid Id { get; set; }
+
+    public DeliveryApiItemDetails? ItemDetails { get; set; }
+}
+
+public class DeliveryApiItemDetails
+{
+    public string? Name { get; set; }
+
+    public IApiContentRoute? Route { get; set; }
+}
+```
+{% endcode %}
+
+The Implementation of the `IDeliveryApiPropertyValueConverter` interface can be found in the following methods:
+
+* `GetDeliveryApiPropertyCacheLevel()`: This method specifies the cache level used for our property representation in the Delivery API response.
+* `GetDeliveryApiPropertyValueType()`: This method defines the value type of the custom property for the Delivery API response. The value type for the API may differ from the type used for Razor rendering _(as specified by `GetPropertyValueType()`)_.
```

Comment:
```suggestion
* `GetDeliveryApiPropertyValueType()`: This method defines the value type of the custom property output for the Delivery API response.
```

Everything can differ from the Razor rendering - that's the whole point. Is there any reason to single out the value type here?

Reply:
You are absolutely right! There wasn't really any particular reason for it, just thought it is an easy-to-relate example 😁